contributors/design-proposals/device-plugin.md

```diff
@@ -0,0 +1,794 @@
+is performing:
+* In general we expect Kubelet to remove any devices that are owned by the failed
+  device plugin from the resources advertised by the Node status.
+* We however do not expect Kubelet to fail or restart any pods or containers
+  running that are using these devices.
+* If Kubelet is in the process of allocating a device, then it should fail
+  the container process and reschedule the Pod.
+
+If the Kubelet fails or restarts, we expect the Device Plugins to know about
+it through Kubelet's Heartbeat call which every Device Plugin should call
+every 5s.
+
+When Kubelet fails or restarts it should know what are the devices that are
+owned by the different containers and be able to rebuild a list of available
+devices.
+In the current design, instead of checkpointing this data, we propose to save
+this in the API server as this gives introspection capabilities to the user
+has minimal impact on performances and is a minimal change that can be
+reverted if we decide to implement checkpointing or a debug API later.
+
+If Kubelet failed and recovered between two Heartbeat we are expecting it
+to answer with a HeartbeatKo answer. Signaling the device plugins to register
+themselves again against the Kubelet (in case of heartbeat failure
+or connection error).
+
+### API Changes
+
+When discovering the devices, Kubelet will be in charge of advertising those
+resources to the API server as part of the kubelet node update current protocol.
+
+We will advertise each device returned by the Device Plugin in a new structure
+called `Device`.
+It is defined as follows:
+
+```golang
+// E.g:
+// struct Device {
+//    Kind: "NVIDIA-gpu"
+//    Name: "GPU-fef8089b-4820-abfc-e83e-94318197576e"
+//    Health: "Healthy",
+//    Properties: {
+//        "Family": "Pascal",
+//        "Memory": "4G",
+//        "ECC"   : "True",
+//    }
+//}
+type Device struct {
+	Kind       string
+	Vendor     string
+	Name       string
+	Health     DeviceHealthStatus
+	Properties map[string]string
+}
+```
+
+Because the current API (Capacity) can not be extended to support Device,
+we will need to create one new attribute in the NodeStatus structure:
+  * `DevCapacity`: Describing the device capacity of the node
+
+```golang
+type NodeStatus struct {
+	DevCapacity []Device
+}
+```
+
+We also introduce the `Devices` field in the Containers status so that user
+can know what devices were assigned to the container.
+
+```golang
+type ContainerStatus struct {
+	Devices []Device
+}
+```
+
+Note that we will be using OIR to schedule and trigger the device plugin
+in parallel.
+So when a Device plugin registers two `foo-device` the node status will be
+updated to advertise 2 `extensions.kubernetes.io/foo-device`.
+
+If a user wants to trigger the device plugin he only needs to request this
+OIR in his Pod Spec.
+
+## Upgrading your cluster
+
+TLDR: If you are upgrading either Kubelet or any device plugin the safest way
```

Comment:
Why is this the recommendation? What if the device plugin has a trivial bug that can be fixed without having to drain?
I think it's better to list recipes for various kinds of upgrades that recommending a one stop solution.